testing/unittests/loggingtests.cpp

```diff
@@ -0,0 +1,485 @@
+        hasServiceGroupPredicate("hasService-id-predicate", LGSTGetTransactionID, "baz", 0, 5);
+    }
+
+    void testHasService()
+    {
+        hasServiceGroup("hasService-seed", LGSTGetTransactionSeed, "", true);
+        hasServiceGroup("hasService-seed", LGSTGetTransactionSeed, "foo", true);
+        hasServiceGroup("hasService-seed", LGSTGetTransactionSeed, "bar", true);
+        hasServiceGroup("hasService-seed", LGSTGetTransactionSeed, "baz", false);
+        hasServiceGroup("hasService-update", LGSTUpdateLOG, "", true);
+        hasServiceGroup("hasService-update", LGSTUpdateLOG, "foo", true);
+        hasServiceGroup("hasService-update", LGSTUpdateLOG, "bar", true);
+        hasServiceGroup("hasService-update", LGSTUpdateLOG, "baz", false);
+        hasServiceGroup("hasService-id", LGSTGetTransactionID, "", true);
+        hasServiceGroup("hasService-id", LGSTGetTransactionID, "foo", true);
+        hasServiceGroup("hasService-id", LGSTGetTransactionID, "bar", true);
+        hasServiceGroup("hasService-id", LGSTGetTransactionID, "baz", false);
+    }
+
+    void testGetTransactionSeed()
+    {
+        getTransactionSeed("getTransactionSeed", "", true, "alpha", "Transaction Seed");
+        getTransactionSeed("getTransactionSeed", "foo", true, "beta", "Transaction Seed");
+        getTransactionSeed("getTransactionSeed", "bar", true, "gamma", "Transaction Seed");
+        getTransactionSeed("getTransactionSeed", "baz", false, "Seed", "Failed");
+    }
+
+    void testGetTransactionId()
+    {
+        getTransactionId("getTransactionId", "", true, "alpha1", "");
+        getTransactionId("getTransactionId", "foo", true, "beta1", "");
+        getTransactionId("getTransactionId", "bar", true, "gamma1", "");
+        getTransactionId("getTransactionId", "baz", false, "", "");
+    }
+
+    LoggingPredicateTests()
+        : m_loader(nullptr, nullptr, nullptr, nullptr)
+    {
+    }
+
+    ILoggingManager* createManager()
+    {
+        Owned<IPTree> config(createPTreeFromXMLString(managerText));
+        Owned<ILoggingManager> manager(m_loader.create(*config));
+
+        manager->init(config, "unittest");
+        return manager.getClear();
+    }
+
+    void hasServiceGroupPredicate(const char* test, LOGServiceType service, const char* group, unsigned expectedPass, unsigned expectedFail)
+    {
+        using Variants = std::vector<Linked<const IEspLogAgentVariant> >;
+        Variants pass;
+        Variants fail;
+        VariantGroupPredicate groupPred(group);
+        EspLogAgentVariantPredicate wrapper = [&](const IEspLogAgentVariant& variant) {
+            if (groupPred(variant))
+                pass.emplace_back(&variant);
+            else
+                fail.emplace_back(&variant);
+            return false;
+        };
+
+        if (!group)
+            group = "";
+
+        Owned<ILoggingManager> manager(createManager());
```

Comment:
Thought about it and opted against. I don't think object reuse between tests is usually recommended, because you don't know if a previous test has changed the state in an unexpected way.

Reply:
I agree, because this is a unit test, each test should be started clean.